File: libs/eave-stdlib-py/src/eave/stdlib/middleware/development_bypass.py
from asgiref.typing import HTTPScope

from ..api_util import get_header_value
from ..config import SHARED_CONFIG
from ..headers import EAVE_DEV_BYPASS_HEADER

"""
## middleware check bypass

To bypass the signature and auth verification middlewares in development mode, the following conditions must be met:

1. `EAVE_ENV` must be set to "development"
1. Python "dev_mode" must be enabled. You can set `PYTHONDEVMODE=1` in your `.env` file to enable it.
1. The `GOOGLE_CLOUD_PROJECT` must not be set to `eave-production`.
1. A header `X-Google-EAVEDEV` is set to "1"

### Auth Bypass

When bypassing auth, the `Authorization` header should contain the ID of the account you want to act as.
"""


def development_bypass_allowed(scope: HTTPScope) -> bool:
    if not SHARED_CONFIG.is_development:
        return False
    if not SHARED_CONFIG.dev_mode:
        return False
    if SHARED_CONFIG.google_cloud_project == "eave-production":
        return False

    dev_header = get_header_value(scope=scope, name=EAVE_DEV_BYPASS_HEADER)
    if not dev_header:
        return False

    if dev_header == "1":
        return True

    raise Exception("development bypass failed")

File: libs/eave-stdlib-ts/src/middleware/development-bypass.ts
import { Request, Response } from "express";
import { constants as httpConstants } from "node:http2";
import { sharedConfig } from "../config.js";
import { EAVE_DEV_BYPASS_HEADER } from "../headers.js";
import { LogContext, eaveLogger } from "../logging.js";

export function developmentBypassAllowed(
  req: Request,
  _res: Response,
): boolean {
  if (!sharedConfig.isDevelopment) {
    return false;
  }
  if (!sharedConfig.devMode) {
    return false;
  }
  if (sharedConfig.googleCloudProject === "eave-production") {
    return false;
  }

  const devHeader = req.header(EAVE_DEV_BYPASS_HEADER);
  if (!devHeader) {
    return false;
  }

  if (devHeader === "1") {
    return true;
  }

  throw new Error("development bypass failed");
}

export function developmentBypassAuth(req: Request, res: Response): void {
  const ctx = LogContext.load(res);
  eaveLogger.warning("Bypassing auth verification in dev env", ctx);

  const accountId = req.header(httpConstants.HTTP2_HEADER_AUTHORIZATION);
  if (accountId === undefined || typeof accountId !== "string") {
    throw new Error("Authorization header was empty");
  }

  // no account lookup since we cant access orm from ts...

  ctx.eave_account_id = accountId;
}
